### ISSUE TITLE
Convolution Op: are both Channel In & Out multiple of group number?

### ISSUE THREAD

**Issue by AlexandreEichenberger** [onnx/onnx#3641]
## Question

Are both the Channel In (C) and Channel Out (M) multiple of the group number for the Conv Op? In the doc, Channel In is specified as being a multiple of the group number. Nothing is said about Channel Out.

## Question Details

Here is what the doc say:
> W (differentiable) : T [...]  X.shape[1] == (W.shape[1] * group) == C
https://github.com/onnx/onnx/blob/master/docs/Operators.md#Conv

Both Pytorch and Keras are specific, indicating that both channel in & out are to be multiple of the number of groups, for grouped convolutions.

> Pytorch requires both channel in (CI) and channel out (CO) to be multiple of group number (G).
https://pytorch.org/docs/stable/generated/torch.nn.Conv2d.html

> Keras also specifies it: Input channels and filters must both be divisible by groups.
https://www.tensorflow.org/api_docs/python/tf/keras/layers/Conv2D

### Further information

Answer to this matters as code needs to be different if supporting channel out (M) that are non-multiple of the number of groups.

I propose that we add the clarification that both are multiple of group number. I can make the change if there is agreement over this.

**Comment by AlexandreEichenberger** [comment]
Conversation on slack, reported here for completeness.

>G. Ramalingam  5 days ago
My guess is that this is just a limitation of the documentation. I don't know enough about the semantics of a grouped convolution, but I guess the shape-inference function should indicate what the output dimensions are expected to be. But the documentation in the op itself is quite limited. It would be great if someone knowledgeable about the op could expand on the documentation.

>Alexandre Eichenberger  4 days ago
@G. Ramalingam thanks for your response. I added this constraint in our onnx-mlir code and thus I will know empirically if someone uses models where this does not hold. We run subsets of the model zoo, so I will know for sure if there is a case that break this assert. I will submit a PR to fix the doc if this holds true for a while.

**Comment by kraiskil** [comment]
Reading the current ONNX documentation there doesn't seem to be any reason to require output feature maps being a multiple of groups. The only place where groups is documented to be used is in sharing the weights tensor on the input channel axis. I.e.`W[M][C/groups][...]`, so the number of feature maps(output channels) is independent of the number of groups.

Having said that, the reason I found this issue is that my onnx backend is currently failing all the Conv tests where `group > 1 ` :) And I am struggling to understand what exactly Conv is supposed to compute if groups are given... I haven't been able to find the semantics for Conv defined, for ONNX, Keras/TF or pytorch.

Why should the number of outputs be a multiple of groups? And should not the weight tensor W then have a shape of `W[M/groups][C/groups][...]`?

**Comment by AlexandreEichenberger** [comment]
Here is my Jupyter notebook code that I use for debugging convolution. I test correctness by comparing with Pytorch.

Why being multiple? Because groups is a trick to reduce the weight sizes, and it make little sense to have channel outs that would use a partial amount of the reused weights. I really think the ONNX specs should be strengthen to requires multiple of group G for both input and output channels.

I also did not find any code out there, this is why I started coding and verifying agains existing code.

Also in the code I use CO for channel out, and CI for channel in. I use COO and CII for indices that are within a given group g, and Co/CO for the absolute sizes/address that include the offset generated by g. Capital letters are for sizes, small caps are for the current indices. 

`my_conv2D` is a simpler version with "if" to remove out of bound accesses, `my_conv2D_v2` is the better version that calculate loop bounds to only iterate over inbound loop iterations.


Hope this helps.

``` python
import numpy as np
import math
import torch

# All parameters should be changed here, they will then propagate naturally where needed.
(N, G, CI, CO, HI, WI, KH, KW)= (1, 3, 9, 6, 53, 57, 3, 3) 
(s, d, p) = (3, 2, 1)
assert CI % G == 0, "expected CI to be a multiple of G"
assert CO % G == 0, "expected CO to be a multiple of G"
# NOTE
# * CI is divided in G chunks of CII size. 
#   ciis index withing group, cig is the index in g taking group into account, and ci = cig + cii
# * CO is also divided in G chuncs of COO size. 
#   coo is index within group, cog is the inded in g taking group into account, and co = cig +coo
# * When G=1, CI==CII, CO==COO, cig=cog=0, ci=cii, co=coo

# Input variables initialized here with the right sizes using above defs.
X = np.random.rand(N, CI, HI, WI)
W = np.random.rand(CO, int(CI/G), KH, KW)

XX = torch.from_numpy(X)
WW = torch.from_numpy(W)
nn = torch.nn.Conv2d(CI, CO, KH, padding=[p, p], stride=[s,s], dilation=[d,d], groups=G, bias=False)
nn.weight.data = WW
YY = nn(XX)
Yref = YY.detach().numpy()
print(np.shape(Yref))

def my_floor(a, b):
    return int(math.floor((1.0*a)/(1.0*b)))
def my_ceil(a, b):
    return int(math.ceil((1.0*a)/(1.0*b)))

# ONNXOpHelper.cpp:getConvDimMap 
# agrees with https://oneapi-src.github.io/oneDNN/dev_guide_convolution.html
#    where their d is one less than ours (ours: default 1; theirs, default 0)
# agrees with https://pytorch.org/docs/stable/generated/torch.nn.Conv2d.html
def my_cond2D_spatial_size(I, p1, p2, K, d, s):
    p = p1+p2
    res = my_floor(I+p-(K-1)*d -1, s) + 1
    return int(res)

def my_conv2D_output_shape(X, W, pad, stride, dilation):
    (N, CI, HI, WI) = np.shape(X)
    (CO, CIperG, KH, KW) = np.shape(W)
    HO = my_cond2D_spatial_size(HI, pad[0], pad[2], KH, dilation[0], stride[0])
    WO = my_cond2D_spatial_size(WI, pad[1], pad[3], KW, dilation[1], stride[1])
    return (N, CO, HO, WO)

def val4(array, i0, i1, i2, i3):
    shape = np.shape(array)
    assert i0>=0 and i0<shape[0], "assert failed in first dim: 0<="+str(i0)+"<"+str(shape[0])
    assert i1>=0 and i1<shape[1], "assert failed in second dim: 0<="+str(i1)+"<"+str(shape[1])
    assert i2>=0 and i2<shape[2], "assert failed in third dim: 0<="+str(i2)+"<"+str(shape[2])
    assert i3>=0 and i3<shape[3], "assert failed in fourth dim: 0<="+str(i3)+"<"+str(shape[3])
    return array[i0, i1, i2, i3]
    
def my_conv2D(X, W, pad, stride, dilation, groups):
    output_shape = my_conv2D_output_shape(X, W, pad, stride, dilation)
    print("output shape is", output_shape)
    res = np.zeros(output_shape)
    (N, CI, HI, WI) = np.shape(X)
    (CO, CII, KH, KW) = np.shape(W)
    (N, CO, HO, WO) = np.shape(res)
    (G, COO) = (groups, int(CO/groups))
    assert CI % G == 0, "expected CI to be a multiple of G"
    assert CO % G == 0, "expected CO to be a multiple of G"
    (ph, pw) = pad[0:2]
    (sh, sw) = stride
    (dh, dw) = dilation
    for n in range(N): # reuse W(*), parallel
        for g in range(G): # reuse X. parallel
            cog = g * COO
            cig = g * CII
            for coo in range(COO): # reuse X, parallel
                co = cog + coo
                for ho in range(HO): #spatial X, W
                    for wo in range(WO): #spatial X, W
                        r = 0.0 # for res[n, co, ho, wo]
                        for cii in range(CII):  
                            ci = cig + cii
                            for kh in range(KH): #spatial X, W
                                for kw in range(KW): #spatial X, V
                                    if (ho * sh + kh * dh - ph >= 0 and ho * sh + kh * dh - ph < HI and \
                                        wo * sw + kw * dw - pw >= 0 and wo * sw + kw * dw - pw < WI):
                                        r = r + val4(X, n, ci, ho * sh + kh * dh - ph, wo * sw + kw * dw - pw) * \
                                            val4(W, co, cii, kh, kw)
                        res[n, co, ho, wo] = r
    return res

Yalex = my_conv2D(X, W, pad=[p,p,p,p], stride=[s,s], dilation=[d,d], groups=G)
print(np.shape(Yalex))
res = np.allclose(Yref, Yalex)
print("both computations are the same?", res)

def my_k_range(o, I, K, p, s, d):
    # translate ho * sh + kh * dh - ph>=0  and ho * sh + kh * dh - ph<HI into min/max for kh
    min_in_range = my_ceil(p - o * s, d)
    max_in_range = my_ceil(I + p - o * s, d)
    return (max(min_in_range, 0), min(max_in_range, K))

def my_conv2D_v2(X, W, pad, stride, dilation, groups):
    output_shape = my_conv2D_output_shape(X, W, pad, stride, dilation)
    res = np.zeros(output_shape)
    (N, CI, HI, WI) = np.shape(X)
    (CO, CII, KH, KW) = np.shape(W)
    (N, CO, HO, WO) = np.shape(res)
    (G, COO) = (groups, int(CO/groups))
    assert CI % G == 0, "expected CI to be a multiple of G"
    assert CO % G == 0, "expected CO to be a multiple of G"
    (ph, pw) = pad[0:2]
    (sh, sw) = stride
    (dh, dw) = dilation
    for n in range(N):
        for g in range(G):
            cog = g * COO
            cig = g * CII
            for coo in range(COO):
                co = cog + coo
                for ho in range(HO):
                    for wo in range(WO):
                        r = 0.0 # for res[n, co, ho, wo]
                        (min_kh, max_kh) = my_k_range(ho, HI, KH, ph, sh, dh)
                        (min_kw, max_kw) = my_k_range(wo, WI, KW, pw, sw, dw)
                        for cii in range(CII):
                            ci = cig + cii
                            for kh in range(min_kh, max_kh):
                                for kw in range(min_kw, max_kw):
                                    r = r + val4(X, n, ci, ho * sh + kh * dh - ph, wo * sw + kw * dw - pw) * \
                                        val4(W, co, cii, kh, kw)
                        res[n, co, ho, wo] = r
    return res

Yalex2 = my_conv2D_v2(X, W, pad=[p,p,p,p], stride=[s,s], dilation=[d,d], groups=G)
print(np.shape(Yalex2))
res2 = np.allclose(Yref, Yalex2)
print("both computations are the same?", res2)
```

**Comment by kraiskil** [comment]
Thanks @AlexandreEichenberger, this really cleared it up for me, and the 'group>1' tests nolonger fail.

What I had not understood from the documentation was:
-  convolutions are independent between groups. Also input and output channels are divided into groups, and input channel N affects only the output channel group N. The input and output groups need not be of similar size.
- Each input channel within a group has its own weight, and these weights are shared over groups.  I.e. one weight is not shared for all channels within the group. (in hindsight this is clear: the 2nd dimension size of the weights is C/g, not g)

Now I see the reasoning behind the idea of having output channels a multiple of the group size. Not strictly necessary, but if all front ends out there have this enforced, it would make sense to have the same restriction in ONNX too. If nothing else, it makes the looping in the back end logic a bit less convoluted. 

I now realize that with the fix for groups I just pushed to onnx2c, it now implicitly expects this restriction... So yes, ONNX needs either this restriction, or a back end test where it does not hold. My vote for the former.

**Comment by AlexandreEichenberger** [comment]
Sure, should I make a PR with the added restriction? 
Glad my code was of help.

**Comment by kraiskil** [comment]
Please do. PR issue tend to get more attention than question issues :) (I'm not an ONNX insider, and I am not familiar with the process used in this project).

BTW, I tried making a `M != N*groups` convolution using the sclblonnx wrapper. It would save an ONNX file without problem, and e.g. netron opens that file without warnings. But running inference over that with onnx-runtime fails.
Code here: https://gist.github.com/kraiskil/5885bf317e345fe61ddf27feb2e4052b

I would guess if adding the restriction to the documentation/specification, it should be added to some ONNX internal sanity check too?
